terraformModuleConstitution/general_rules.md

```diff
@@ -0,0 +1,123 @@
+
+## Purpose of Modules
+
+A module should focus on the best practice of a specific service under a certain usage scenario. The resources in the module should be highly cohesive, in the meantime they should be designed to use with other modules as well.
+
+Modules should have consistent style with the official documentation of involved `resource`, `data`, as well as other modules following these rules. Consistent user experience should be provided at all time to reduce the risk of mistakes and misunderstandings.
+
+## Types of Modules
+
+Based on the purpose, there are 2 types of modules.
+
+1. Modules to create real `resource`
+2. Modules contain only `variable`, `locals` and `output`
+
+The modules mentioned below are the first type if not specified.
+
+## Design of Modules
+
+A module should refer to the concept of aggregated root in DDD (Domain Driven Development), there will be a dominated resource and a series of resources work with it.
+
+The information aggregated root resource depends on should be passed through corresponding `variable` instead of creating on its own or inquiring with `data`. [HashiCorp defined dependency-inversion](https://www.terraform.io/docs/language/modules/develop/composition.html#dependency-inversion) should be practiced.
+
+## Consist of a Module
+
+```config
+.
+├── LICENSE
+├── Makefile
+├── README.md
+├── README.yaml
+├── CHANGE.log
+├── context.tf
+├── docs
+│   ├── targets.md
+│   └── terraform.md
+├── examples
+│   └── startup
+│       ├── context.tf
+│       ├── fixtures.us-east-2.tfvars
+│       ├── main.tf
+│       ├── outputs.tf
+│       ├── variables.tf
+│       └── versions.tf
+├── locals.tf
+├── main.tf
+├── outputs.tf
+├── test
+│   ├── Makefile
+│   ├── Makefile.alpine
+│   ├── e2e
+│   ├── go.mod
+│   ├── go.sum
+│   ├── upgrade
+|   └── unit_test
+├── variables.tf
+└── versions.tf
+```
+
+This is a typical structure of a module project.
+
+A Module consists of an author written part and an auto generated part.
+
+The below parts should be written by the author:
+
+* Logical code of the module - all `.tf` files under the root directory
+* File used to generate documentations - README.yaml
+* Automated tests - `e2e`, `upgrade` and `unit_test` directory under `test` directory
+* Examples - `example` directory
+
+The rest parts are rely on code auto generation. The author should not modify the content of these auto generated files. The content will be generated automatically before every commit and overwrite the previous contents.
+
+We consider these parts equally important, so they should have the same code quality standard.
+
+## Static Files and Template Files
+
+Under some scenarios the static file and template files the module will use are saved under `files` directory (Eg. a long script file configured to [`user_data`](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/linux_virtual_machine#user_data)).
```

Comment:
My bad, I forgot to remove `template files` after adding `template` rule. Better be:

```markdown
Under some scenarios the static files that would be used by the module are saved under `files` directory...
```